beacon_node/lighthouse_network/src/peer_manager/mod.rs

```diff
@@ -1087,146 +1333,54 @@ impl<E: EthSpec> PeerManager<E> {
-                            if info.is_outbound_only()
-                                && self.target_outbound_peers()
-                                    >= connected_outbound_peer_count
-                                        .saturating_sub(outbound_peers_pruned)
-                            {
-                                // Restart the main loop with the outbound peer removed from
-                                // the list. This will lower the peers per subnet count and
-                                // potentially a new subnet may be chosen to remove peers. This
-                                // can occur recursively until we have no peers left to choose
-                                // from.
-                                continue;
-                            }
-
-                            // Check the sync committee
-                            if let Some(subnets) = peer_to_sync_committee.get(candidate_peer) {
-                                // The peer is subscribed to some long-lived sync-committees
-                                // Of all the subnets this peer is subscribed too, the minimum
-                                // peer count of all of them is min_subnet_count
-                                if let Some(min_subnet_count) = subnets
-                                    .iter()
-                                    .filter_map(|v| sync_committee_peer_count.get(v).copied())
-                                    .min()
-                                {
-                                    // If the minimum count is our target or lower, we
-                                    // shouldn't remove this peer, because it drops us lower
-                                    // than our target
-                                    if min_subnet_count <= MIN_SYNC_COMMITTEE_PEERS {
-                                        // Do not drop this peer in this pruning interval
-                                        continue;
-                                    }
-                                }
-                            }
-
-                            if info.is_outbound_only() {
-                                outbound_peers_pruned += 1;
-                            }
-                            // This peer is suitable to be pruned
-                            removed_peer_index = Some(index);
-                            break;
+                    .map(|(subnet_id, _)| *subnet_id);
+
+                if let Some(densest_subnet) = custody_subnet_with_most_peers {
+                    // If we have successfully found a candidate peer to prune, prune it,
+                    // otherwise all peers on this subnet should not be removed due to our
+                    // outbound limit or min_subnet_count. In this case, we remove all
+                    // peers from the pruning logic and try another subnet.
+                    if let Some(candidate_peer) = self.find_prune_candidate(
```

Comment:
I think there is a bug here. 

We have built a set of `peers_to_prune`. But I dont think these peers are being accounted for when counting the peers on the columns. 

For example; if (because of scoring) we prune a column down to 1 peer, I think potentially we could still prune that last peer as a candidate, because we've not accounted for the peers that we are about to prune. I think we need to parse the `peers_to_prune` mapping into this and subsequent functions to ensure we're dynamically accounting for peers being pruned as we go.

Reply:
We are removing the peer from the `custody_subnet_to_peers` map and passing that down to `find_prune_candidate()` and `should_protect_peer`, and the updated map is used when counting the peers, is this what you meant?

Reply:
https://github.com/sigp/lighthouse/blob/4a311980d0e9ba33a1223a1af03eb15dc0b2cea2/beacon_node/lighthouse_network/src/peer_manager/mod.rs#L1372-L1374

Reply:
@ackintosh found a related bug with another map - I'm fixing it now.

Reply:
Fixed in https://github.com/sigp/lighthouse/pull/7915/commits/20483f9feb7594c2d44b3972cdc49d295e66d014